### PR TITLE
[GITHUB-2897] Not exception but warning if some (not all) of the give…

### LINKED ISSUE #2897: [Feature] Not exception but warning if some (not all) of the given test names are not found in suite files.
Now, when any of the test names are not found in suite files, TestNG will throw exception.
`The test(s) <[TestA, TestB, TestC]> cannot be found in suite.`

It's reasonable but not so friendly still. For example, if user provide 5 test names and 2 of them cannot be found in the suite files, other 3 tests could be still run.
In this case, it will give user good experience if TestNG print warning message to tell which tests cannot be found and print info message to tell which tests have been found and will be executed.

What do you think?

### PR DESCRIPTION
**Pull request by wenijinew**
…n test names are not found in suite files.

Change-Id: Ica210f6c05b395f1d0a6718479f71b645ebd39cc

Fixes #2897 .

### Did you remember to?

- [x] Add test case(s)
- [x] Update `CHANGES.txt`
- [x] Auto applied styling via `./gradlew autostyleApply`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

**Note:** For more information on contribution guidelines  please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

### CODE REVIEW

File: testng-core/src/test/java/org/testng/xml/internal/TestNamesMatcherTest.java [hunk 1]
```diff
@@ -90,6 +90,38 @@ public void testCloneIfContainsTestsWithNamesMatchingAnyWithoutMatch() {
...
+        new TestNamesMatcher(xmlSuite, Collections.singletonList("test3"));
+    testNamesMatcher.validateMissMatchedTestNames(ignoreMissedTestNames);
+  }
+
+  @Test(
+      description = "GITHUB-2897, Expected exception thrown when ignoreMissedTestNames disabled.",
+      expectedExceptions = TestNGException.class,
+      expectedExceptionsMessageRegExp =
+          "\nThe test\\(s\\) \\<\\[test3\\]\\> cannot be found in suite.")
+  public void testNoExceptionFromValidateWhenIgnoreMissedTestNamesDisabled() {
```
Comment (wenijinew) [inline testng-core/src/test/java/org/testng/xml/internal/TestNamesMatcherTest.java]: NoException -> HaveException

**Review by wenijinew — COMMENTED** [review]
Fixed typos and import issues

File: testng-core/src/test/java/org/testng/xml/internal/TestNamesMatcherTest.java [hunk 2]
```diff
@@ -112,6 +112,30 @@ public void testHaveExceptionFromValidateWhenIgnoreMissedTestNamesDisabled() {
     testNamesMatcher.validateMissMatchedTestNames(ignoreMissedTestNames);
   }
 
+  @Test(description = "GITHUB-2897, No exception thrown when ignoreMissedTestNames enabled by System property 'testng.ignore.missed.testnames'.")
+  public void testNoExceptionFromValidateWhenIgnoreMissedTestNamesEnabledBySystemProperty() {
+    final boolean ignoreMissedTestNames = false;
+    System.setProperty(RuntimeBehavior.TESTNG_IGNORE_MISSED_TESTNAMES, true);
```
Comment (wenijinew) [inline testng-core/src/test/java/org/testng/xml/internal/TestNamesMatcherTest.java:118]: true -> "true"

File: testng-core/src/test/java/org/testng/xml/internal/TestNamesMatcherTest.java [hunk 3]
```diff
@@ -112,6 +112,30 @@ public void testHaveExceptionFromValidateWhenIgnoreMissedTestNamesDisabled() {
...
+  }
+
+  @Test(
+      description = "GITHUB-2897, Expected exception thrown when ignoreMissedTestNames disabled by System property 'testng.ignore.missed.testnames'.",
+      expectedExceptions = TestNGException.class,
+      expectedExceptionsMessageRegExp =
+          "\nThe test\\(s\\) \\<\\[test3\\]\\> cannot be found in suite.")
+  public void testHaveExceptionFromValidateWhenIgnoreMissedTestNamesDisabledBySystemProperty() {
+    final boolean ignoreMissedTestNames = false;
+    System.setProperty(RuntimeBehavior.TESTNG_IGNORE_MISSED_TESTNAMES, false);
```
Comment (wenijinew) [inline testng-core/src/test/java/org/testng/xml/internal/TestNamesMatcherTest.java:132]: false -> "false"

Comment (wenijinew) [reply testng-core/src/test/java/org/testng/xml/internal/TestNamesMatcherTest.java:132]: import RuntimeBehavior

File: testng-core/src/main/java/org/testng/xml/internal/TestNamesMatcher.java [hunk 4]
```diff
@@ -56,7 +56,7 @@ public void validateMissMatchedTestNames(final boolean ignoreMissedTestNames) {
     final List<String> tmpTestNames = getMissedTestNames();
     if (!tmpTestNames.isEmpty()) {
       final String errMsg = "The test(s) <" + tmpTestNames + "> cannot be found in suite.";
-      if (ignoreMissedTestNames) {
+      if (ignoreMissedTestNames || RuntimeBehavior.ignoreMissedTestNames()) {
```
Comment (wenijinew) [inline testng-core/src/main/java/org/testng/xml/internal/TestNamesMatcher.java:59]: import RuntimeBehavior

**Review by wenijinew — COMMENTED** [review]
to fix issues about import and boolean values

File: testng-core/src/main/java/org/testng/JarFileUtils.java [hunk 5]
```diff
@@ -104,15 +104,20 @@ private boolean testngXmlExistsInJar(File jarFile, List<String> classes) throws
...
+        // won't run any test names. (default legacy logic)
         if (testNames != null) {
           TestNamesMatcher testNamesMatcher = new TestNamesMatcher(suite, testNames);
-          testNamesMatcher.validateMissMatchedTestNames(ignoreMissedTestNames);
-          suites.addAll(testNamesMatcher.getSuitesMatchingTestNames());
+          boolean validationResult =
+              testNamesMatcher.validateMissMatchedTestNames(ignoreMissedTestNames);
+          if (validationResult) {
+            suites.addAll(testNamesMatcher.getSuitesMatchingTestNames());
+          }
```
Comment (wenijinew) [inline testng-core/src/main/java/org/testng/JarFileUtils.java:115]: else print warning message to tell there is no test to run?

File: testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java [hunk 6]
```diff
@@ -115,6 +116,11 @@ public static boolean isDryRun() {
     return Boolean.parseBoolean(value);
   }
 
+  public static boolean ignoreMissedTestNames() {
```
Comment (krmahadevan) [inline testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java:119]: I dont think this is required since we are providing a way to set this via configurations. This means that a user should be able to leverage build tools plugins such as surefire plugin and then set this value.  Please have this removed.

File: testng-core/src/main/java/org/testng/xml/internal/TestNamesMatcher.java [hunk 7]
```diff
@@ -43,13 +50,42 @@ public List<XmlSuite> getSuitesMatchingTestNames() {
...
+   * ignoreMissedTestNames is enabled, then any of the given test names exist in suite will be run,
+   * and print warning message to tell those test names do not exist in suite.
+   *
+   * @param ignoreMissedTestNames if true print warning message otherwise throw TestNGException for
+   *     missed testNames.
+   * @return boolean if ignoreMissedTestNames disabled, then return true if no missed test names in
+   *     suite, otherwise throw TestNGException; if ignoreMissedTestNames enabled, then return true
+   *     if any test names exist in suite, otehrwise (all given test names are missed) false.
+   */
+  public boolean validateMissMatchedTestNames(final boolean ignoreMissedTestNames) {
```
Comment (krmahadevan) [inline testng-core/src/main/java/org/testng/xml/internal/TestNamesMatcher.java:67]: I am not sure if we need to be setting up `final` for parameters or for local variables.

File: CHANGES.txt [hunk 8]
```diff
@@ -1,4 +1,5 @@
 Current
+New: GITHUB-2897: Not exception but warning if some (not all) of the given test names are not found in suite files. (Bruce Wen)
```
Comment (krmahadevan) [inline CHANGES.txt:2]: Can we please revert the formatting changes in this file ?

**Comment by krmahadevan** [comment]
@wenijinew -  I see this warning `This branch cannot be rebased due to conflicts` on github on your PR. Can you please help fix this ? 

Also please help add a test that vets out the below behaviour.

1. User has tests `t1, t2,t3` in their suite.
2. User specifies the tests to run as `t4, t5`.
3. User has enabled the newly introduced configuration parameter that would basically ignore missing tests.

What would the behaviour be? Should TestNG return success, even though it didn't run any tests or should TestNG fail ?

File: CHANGES.txt (hunk 8 above)
Comment (wenijinew) [reply CHANGES.txt:2]: Yes, to fix.

File: testng-core/src/main/java/org/testng/xml/internal/TestNamesMatcher.java (hunk 7 above)
Comment (wenijinew) [reply testng-core/src/main/java/org/testng/xml/internal/TestNamesMatcher.java:67]: any concern about using final here?

File: testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java (hunk 6 above)
Comment (wenijinew) [reply testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java:119]: this method is used in other places.

Comment (wenijinew) [reply testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java:119]: And this method is to get the System property, not set.

File: testng-core/src/main/java/org/testng/JarFileUtils.java (hunk 5 above)
Comment (wenijinew) [reply testng-core/src/main/java/org/testng/JarFileUtils.java:115]: if it's false, an exception will be thrown.
false means there is no any test names matching the existing ones, and nothing to run.

**Comment by wenijinew** [comment]
> 

Exception will be thrown if all given test names are invalid. Please help check the latest code change and new test cases. Thanks.

File: testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java (hunk 6 above)
Comment (wenijinew) [reply testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java:119]: I understood now. However, in our case, the values of test names are dynamic and the better way is to use the Vm argument to specify them.

**Comment by wenijinew** [comment]
> @wenijinew - I see this warning `This branch cannot be rebased due to conflicts` on github on your PR. Can you please help fix this ?
> 
> Also please help add a test that vets out the below behaviour.
> 
> 1. User has tests `t1, t2,t3` in their suite.
> 2. User specifies the tests to run as `t4, t5`.
> 3. User has enabled the newly introduced configuration parameter that would basically ignore missing tests.
> 
> What would the behaviour be? Should TestNG return success, even though it didn't run any tests or should TestNG fail ?

I suppose the conflicts are gone now.

**Comment by wenijinew** [comment]
> > 
> 
> Exception will be thrown if all given test names are invalid. Please help check the latest code change and new test cases. Thanks.
> 
> One special case: given test names are empty. -testnames "". I've also made change for that.
> 
> if (testNames != null && !testNames().isEmpty())
> 
> Instead of throwing an strange exception like "The test(s) <[]> cannot be found", the whole suite will be run. Hope you agree on it. https://groups.google.com/g/testng-users/c/bzuyXNtECFw

In that case, maybe more special input need to consider: test names are not an empty list but each element in the list is blank. :(

File: testng-core/src/main/java/org/testng/JarFileUtils.java (hunk 5 above)
Comment (wenijinew) [reply testng-core/src/main/java/org/testng/JarFileUtils.java:115]: false here only means the test names are not found in the current suite.

Comment (wenijinew) [reply testng-core/src/main/java/org/testng/JarFileUtils.java:132]: Print error message and return false here.
However, it's hard to imagine one situation that:
1. given test names are valid: non-null, non-empty, non-all test names are null/empty/blank.
2. the missed test names are empty and matched test names are also mepty. 
According to current logic, it would be never to happen.

File: testng-core/src/test/java/org/testng/JarFileUtilsTest.java [hunk 9]
```diff
@@ -89,11 +126,32 @@ public void testWithInvalidXmlFile() throws MalformedURLException {
...
+  @Test(
+      expectedExceptions = TestNGException.class,
+      expectedExceptionsMessageRegExp = "\nThe test\\(s\\) <\\[dummy\\]> cannot be found in suite.")
+  public void ensureThatExceptionAreThrown() throws MalformedURLException {
     TestNG testNg = new TestNG(false);
     List<String> testNames =
         Arrays.asList("testng-tests-child2", "testng-tests-child4", "testng-tests-child5", "dummy");
     testNg.setTestNames(testNames);
-    testNg.setXmlPathInJar(jar.getAbsolutePath());
+    testNg.setXmlPathInJar("jarfileutils/testng-tests.xml");
```
Comment (wenijinew) [inline testng-core/src/test/java/org/testng/JarFileUtilsTest.java:154]: This is the fix for legacy test case issue: used wrong value for xml path in jar.

File: testng-core/src/main/java/org/testng/xml/internal/TestNamesMatcher.java (hunk 7 above)
Comment (wenijinew) [reply testng-core/src/main/java/org/testng/xml/internal/TestNamesMatcher.java]: The parameter is removed in new push. Instead, use the field and intialize in constructor. Therefore, the value should come from TestNG by option or VM argument.

**Comment by wenijinew** [comment]
> Also please help add a test that vets out the below behaviour.
> 
> 1. User has tests `t1, t2,t3` in their suite.
> 2. User specifies the tests to run as `t4, t5`.
> 3. User has enabled the newly introduced configuration parameter that would basically ignore missing tests.
> 
> What would the behaviour be? Should TestNG return success, even though it didn't run any tests or should TestNG fail ?

TestNamesMatcherTest#testHaveExceptionWhenIgnoreMissedTestNamesEnabledWithAllInvalidTestNames added for this and TestNGException is expected.

**Comment by wenijinew** [comment]
> One special case: given test names are empty. -testnames "". I've also made change for that.
> 
> if (testNames != null && !testNames().isEmpty())
> 
> Instead of throwing an strange exception like "The test(s) <[]> cannot be found", the whole suite will be run. Hope you agree on it. https://groups.google.com/g/testng-users/c/bzuyXNtECFw

Any null or empty list, or any list with all elements are null or empty or blank are treated as invalid input, and exception will be thrown from TestNamesMatcher: 

`throw new TestNGException("Please provide a valid list of names to check.")`

**Comment by wenijinew** [comment]
@krmahadevan 

Could you please help review the latest commit? Thanks!

File: testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java (hunk 6 above)
Comment (krmahadevan) [reply testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java:119]: @wenijinew - I think the command line arguments should work without having the need for JVM arguments no? We are trying to move away from having to introduce so many JVM arguments and would like to stick to command line arguments that can be set either via `java org.testng.TestNG` (or) via a build tool. I dont think we need a redundant JVM argument way of doing it.

Comment (wenijinew) [reply testng-core-api/src/main/java/org/testng/internal/RuntimeBehavior.java:119]: OK, I understand. It's not required but better to have to some degree. So, I will remove VM arg in new commit.

**Comment by wenijinew** [comment]
@krmahadevan 

As requested, the VM argument has been removed in the latest commit. Please help review. Thanks.

**Comment by krmahadevan** [comment]
@wenijinew - I am again seeing the same `This branch cannot be rebased due to conflicts` error msg. Can you please help fix this ?

**Review by krmahadevan — APPROVED** [review]

**Comment by krmahadevan** [comment]
@juherr - Please help take a quick look. We would need to squash and merge this PR, because there are a lot of commits in this PR.

**Review by juherr — APPROVED** [review]
